### PR TITLE
PR: Watch a project file tree for changes

### PR DESCRIPTION
**Pull request by andfoy**
## Description of Changes
Added a watchdog observer to receive notifications about files and folders changes. i.e., Additions, deletions, move to other folders.

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)

* [ ] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))

### Issue(s) Resolved

Required by #9482 

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

I certify the above statement is true and correct:
@andfoy

### CODE REVIEW

File: spyder/plugins/projects/workspace/watcher.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        logger.info("Created {0}: {1}".format(
+            self.fmt_is_dir(is_dir), src_path))
+        self.sig_file_created.emit(src_path, is_dir)
+
+    def on_deleted(self, event):
+        src_path = event.src_path
+        is_dir = event.is_directory
+        logger.info("Deleted {0}: {1}".format(
+            self.fmt_is_dir(is_dir), src_path))
+        self.sig_file_created.emit(src_path, is_dir)
```
Comment (ccordoba12) [inline spyder/plugins/projects/workspace/watcher.py]: I think this should be `sig_file_deleted`.

File: spyder/plugins/projects/workspace/watcher.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+    def on_modified(self, event):
+        src_path = event.src_path
+        is_dir = event.is_directory
+        logger.info("Modified {0}: {1}".format(
+            self.fmt_is_dir(is_dir), src_path))
+        self.sig_file_created.emit(src_path, is_dir)
```
Comment (ccordoba12) [inline spyder/plugins/projects/workspace/watcher.py]: And this `sig_file_modified`.

File: spyder/plugins/projects/workspace/watcher.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
```
Comment (ccordoba12) [inline spyder/plugins/projects/workspace/watcher.py]: Please add a docstring to this module.

File: spyder/plugins/projects/workspace/watcher.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+# Third-party imports
+from qtpy.QtCore import QObject, Signal, Slot
+
+from watchdog.observers import Observer
+from watchdog.events import FileSystemEventHandler
+
+logger = logging.getLogger(__name__)
+
+
+class WorkspaceEventHandler(QObject, FileSystemEventHandler):
```
Comment (ccordoba12) [inline spyder/plugins/projects/workspace/watcher.py]: Please add a docstring for this class.

File: spyder/plugins/projects/workspace/watcher.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+    def on_modified(self, event):
+        src_path = event.src_path
+        is_dir = event.is_directory
+        logger.info("Modified {0}: {1}".format(
+            self.fmt_is_dir(is_dir), src_path))
+        self.sig_file_created.emit(src_path, is_dir)
+
+
+class WorkspaceWatcher(QObject):
```
Comment (ccordoba12) [inline spyder/plugins/projects/workspace/watcher.py]: And also a docstring for this class.

**Comment by andfoy** [comment]
> That way this watcher will be started both in the File and Project Explorers.

Maybe we should run some tests on a large directory to see if watchdog does not crash, at least that would be the case when Spyder is started on the home directory.

**Comment by ccordoba12** [comment]
> Maybe we should run some tests on a large directory to see if watchdog does not crash, at least that would be the case when Spyder is started on the home directory.

The idea is to have activate the watcher only a project is active. Otherwise, it shouldn't be necessary.

**Comment by andfoy** [comment]
>  I think you should move this to spyder/workers and add it to the File explorer directly

The idea here is to update the file explorer when watchdog notifies a change?

**Comment by goanpeca** [comment]
@andfoy currently the file explorer is getting all these updates, no?? how is it working at the moment?

**Comment by ccordoba12** [comment]
> The idea here is to update the file explorer when watchdog notifies a change?

Yeah, but only to the root folder (or folders) of a project (or projects).

> currently the file explorer is getting all these updates, no?? how is it working at the moment?

Nothing is connected yet.

**Comment by goanpeca** [comment]
>Nothing is connected yet.

I meant that if I delete a file on the OS the file explorer is updated. So there must be some detection going on ?

**Comment by andfoy** [comment]
> I meant that if I delete a file on the OS the file explorer is updated. So there must be some detection going on ?

I think the tree updates when the file is deleted/renamed directly on Spyder, what happens if the file is removed outside Spyder?

**Comment by goanpeca** [comment]
file

**Comment by andfoy** [comment]
> It works... (it has been like this for a loooong time)

https://github.com/spyder-ide/spyder/blob/f0e625450c6fb74347eb474989f2cab1c4099977/spyder/plugins/explorer/widgets.py#L1086

However, this won't tell us if a file/directory changes its name

**Comment by ccordoba12** [comment]
That only changes the collapsed state, so it has no effect on the file system, I'd say.

**Comment by andfoy** [comment]
@ccordoba12 So we stick with developing the file watcher?

**Comment by goanpeca** [comment]
I am not saying we should not develop. I was just wondering how the file explorer does it (if it is using a Qt thing or what?) and also the watcher if developed to include missing functionality should allow a way to hook it to the file explorer (and the project explorer)

**Comment by ccordoba12** [comment]
Yeah, I think we should keep developing it because `QFileSystemModel` doesn't seem to have the functionality that we need:

https://doc.qt.io/qt-5/qfilesystemmodel.html#signals

**Comment by ccordoba12** [comment]
However, this could be used instead:

https://doc.qt.io/qt-5/qfilesystemwatcher.html#details

(although, again, it doesn't seem to have everything that we need).

**Comment by andfoy** [comment]
@ccordoba12 This one is blocked by https://github.com/conda-forge/watchdog-feedstock/issues/9

**Comment by ccordoba12** [comment]
> This one is blocked by conda-forge/watchdog-feedstock#9

Solved. I moved the new watchdog 0.9.0 package to our Anaconda channel, so everything should be fine now.

**Comment by andfoy** [comment]
5

**Comment by andfoy** [comment]
@ccordoba12 We're still having issues with watchdog installation on some images

**Comment by ccordoba12** [comment]
Please try again, it should work now.

File: requirements/conda.txt [hunk 6]
```diff
@@ -19,4 +19,4 @@ qdarkstyle >=2.6.4
 atomicwrites
 diff-match-patch
 pexpect
-
+watchdog
```
Comment (ccordoba12) [inline requirements/conda.txt]: Please also add `watchdog` to our `setup.py` and our Readme.

Comment (goanpeca) [reply requirements/conda.txt]: Should we add this also to the `requirements.txt?`

**Comment by andfoy** [comment]
@ccordoba12 Tests run locally, however they are hanging on CIs

File: spyder/plugins/projects/workspace/watcher.py (hunk 3 above)
Comment (ccordoba12) [reply spyder/plugins/projects/workspace/watcher.py]: This module still has no docstring.

File: spyder/plugins/projects/workspace/watcher.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+from watchdog.observers import Observer
+from watchdog.events import FileSystemEventHandler
+
+logger = logging.getLogger(__name__)
+
+
+class WorkspaceEventHandler(QObject, FileSystemEventHandler):
+    """Event handler for watchdog notifications. This class recieves
+       notifications about file/folder moving, modification,
+       creation and deletion and emits its corresponding signal."""
```
Comment (ccordoba12) [inline spyder/plugins/projects/workspace/watcher.py]: This is a very odd way to write docstrings. Please change it too:

```python
"""
Event handler for watchdog notifications.

This class receives notifications about file/folder moving, modification,
creation and deletion and emits a corresponding signal about it.
"""
```

File: spyder/plugins/projects/workspace/watcher.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+        src_path = event.src_path
+        is_dir = event.is_directory
+        logger.info("Modified {0}: {1}".format(
+            self.fmt_is_dir(is_dir), src_path))
+        self.sig_file_modified.emit(src_path, is_dir)
+
+
+class WorkspaceWatcher(QObject):
+    """Wrapper class around watchdog observer and notifier. Provides methods
+       to start and stop watching folders."""
```
Comment (ccordoba12) [inline spyder/plugins/projects/workspace/watcher.py]: This docstring needs to be

```python
"""
Wrapper class around watchdog observer and notifier.

It provides methods to start and stop an observer to watch folders.
"""
```

File: spyder/plugins/projects/tests/test_plugin.py [hunk 9]
```diff
@@ -289,5 +292,95 @@ def test_project_explorer_tree_root(projects, tmpdir, qtbot):
...
+    # Test folder creation
+    with qtbot.waitSignal(fs_handler.sig_file_created,
+                          timeout=3000) as blocker:
+        folder2 = project_root.mkdir('folder2')
+
+    folder_created, is_dir = blocker.args
+    assert folder_created == osp.join(to_text_string(project_root), 'folder2')
+
+    # Test file move/renaming
+    new_file = osp.join(to_text_string(folder0), 'new_file')
```
Comment (ccordoba12) [inline spyder/plugins/projects/tests/test_plugin.py]: I could only make this work if `new_file` is in the same directory as the original file. Is this perhaps a limitation of watchdog?

Comment (andfoy) [reply spyder/plugins/projects/tests/test_plugin.py]: At least on my local setup, it is working as expected, watchdog should be monitoring all the files and folders under the same root folder. I.e., recursive option is set to True

Comment (ccordoba12) [reply spyder/plugins/projects/tests/test_plugin.py]: Ok, that's great to know!

**Comment by andfoy** [comment]
> Besides, I really don't know why the new test fails in macOS

Maybe is due to https://github.com/conda-forge/watchdog-feedstock/pull/8

**Comment by ccordoba12** [comment]
You're most probably right. That package can't be `noarch` (the guy of that PR confused me because he said that the previous `noarch` package was working, but there was no such package).

File: spyder/plugins/projects/tests/test_plugin.py [hunk 10]
```diff
@@ -289,5 +292,95 @@ def test_project_explorer_tree_root(projects, tmpdir, qtbot):
         assert topleft_index.data() == osp.basename(ppath)
 
 
+@pytest.mark.skipif(sys.platform == 'darwin', reason='Fails in macOS')
```
Comment (ccordoba12) [inline spyder/plugins/projects/tests/test_plugin.py]: I uploaded new packages for `watchdog` that include the missing library in macOS, so this shouldn't be necessary.

Please remove it to see what happens. If tests pass then this should be ready to go.

File: spyder/plugins/projects/workspace/watcher.py (hunk 3 above)
Comment (ccordoba12) [reply spyder/plugins/projects/workspace/watcher.py]: And please address this one!

**Comment by andfoy** [comment]
@ccordoba12 It seems that we're still having issues with watchdog on Mac. @goanpeca does the test run in Mac locally?

**Comment by ccordoba12** [comment]
> It seems that we're still having issues with watchdog on Mac

I skipped the failing part of the new test on macOS. I think it's really a minor, so we can skip it for now.

**Comment by goanpeca** [comment]
I will check in a couple of hours

**Comment by ccordoba12** [comment]
Thanks @andfoy, this looks really good!

**Pull request merged**